packages/flutter/lib/src/services/autofill.dart

```diff
@@ -0,0 +1,153 @@
+import 'package:flutter/foundation.dart';
+
+import 'text_input.dart';
+
+@immutable
+class AutofillConfiguration {
+  const AutofillConfiguration({
+    @required this.uniqueIdentifier,
+    @required this.autofillHints,
+    this.currentEditingValue,
+  }) : assert(uniqueIdentifier != null),
+       assert(autofillHints != null);
+
+  final String uniqueIdentifier;
+  final List<String> autofillHints;
+  final TextEditingValue currentEditingValue;
+
+  /// Returns a representation of this object as a JSON object.
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'uniqueIdentifier': uniqueIdentifier,
+      'hints': autofillHints,
+      'editingValue': currentEditingValue.toJSON(),
+    };
+  }
+}
+
+// Ideally AutofillClient should NOT extend TextInputClient. Rather, TextInputClient
+// should have an `AutofillDelegate` property.
+abstract class AutofillClient extends TextInputClient {
+  TextInputConfiguration get textInputConfiguration;
+
+  String get uniqueIdentifier;
+
+  AutofillScope get currentAutofillScope;
+
+  //TextEditingValue get currentTextEditingValue;
+  //void updateEditingValue(TextEditingValue value);
+}
+
+mixin AutofillClientMixin implements AutofillClient {
+  @override
+  String get uniqueIdentifier {
+    final String identifier = textInputConfiguration.autofillConfiguration?.uniqueIdentifier;
+    assert(identifier != null);
+    return identifier;
+  }
+}
+
+abstract class AutofillScope {
+  String get uniqueIdentifier;
+
+  AutofillClient getAutofillClient(String tag);
+
+  // Updating TextInputConfiguration is not needed as far as autofill goes,
+  // because changing focus is the only way to trigger autofill.
+  //void updateTextInputConfigurationIfNeeded();
```

Comment:
Updating the platform side only on focus change is a bit questionable when the autofill service saves the user input. It could cause the wrong input being saved by the autofill service for future autofill.